### PR TITLE
[var] Add a new API, hb_font_set_var_named_instance

### LINKED ISSUE #1864: Selecting a named instance
It would be nice to have an api to select a named instance of a hb_face_t.

### PR DESCRIPTION
**Pull request by ebraminio**
Fixes #1864

### CODE REVIEW

File: test/api/test-ot-extents-cff.c [hunk 1]
```diff
@@ -184,6 +184,37 @@ test_extents_cff2_vsindex (void)
...
+test_extents_cff2_vsindex_named_instance (void)
+{
+  hb_face_t *face = hb_test_open_font_file ("fonts/AdobeVFPrototype_vsindex.otf");
+  g_assert (face);
+  hb_font_t *font = hb_font_create (face);
+  hb_face_destroy (face);
+  g_assert (font);
+  hb_ot_font_set_funcs (font);
+
+  hb_font_set_var_named_instance (font, 6); // 6 (BlackMediumContrast): 900, 50
```
Comment (ebraminio) [inline test/api/test-ot-extents-cff.c]: Checked it manually against { 900, 50 } actual results.

File: src/hb-font.cc [hunk 2]
```diff
@@ -1881,6 +1881,32 @@ hb_font_set_var_coords_design (hb_font_t *font,
...
+hb_font_set_var_named_instance (hb_font_t *font,
+				unsigned instance_index)
+{
+  if (hb_object_is_immutable (font))
+    return;
+
+  unsigned int coords_length = hb_ot_var_named_instance_get_design_coords (font->face, instance_index, nullptr, nullptr);
+  if (!coords_length) return;
+
+  float *coords = (float *) malloc (sizeof (float) * coords_length);
```
Comment (behdad) [inline src/hb-font.cc]: calloc.

File: src/hb-font.cc [hunk 3]
```diff
@@ -1881,6 +1881,32 @@ hb_font_set_var_coords_design (hb_font_t *font,
...
+  if (!coords) return;
```
Comment (behdad) [inline src/hb-font.cc]: unlikely.

File: src/hb-font.cc (hunk 2 above)
Comment (behdad) [reply src/hb-font.cc]: Look at what `hb_font_set_var_coords_design` does:
```
  int *normalized = coords_length ? (int *) calloc (coords_length, sizeof (int)) : nullptr; 
  if (unlikely (coords_length && !normalized)) 
    return; 
```

Comment (ebraminio) [reply src/hb-font.cc]: Done

File: src/hb-font.cc (hunk 3 above)
Comment (ebraminio) [reply src/hb-font.cc]: Done

File: src/hb-font.cc [hunk 4]
```diff
@@ -1881,6 +1882,35 @@ hb_font_set_var_coords_design (hb_font_t *font,
...
+ */
+void
+hb_font_set_var_named_instance (hb_font_t *font,
+				unsigned instance_index)
+{
+  if (hb_object_is_immutable (font))
+    return;
+
+  unsigned int coords_length = hb_ot_var_named_instance_get_design_coords (font->face, instance_index, nullptr, nullptr);
+  if (!coords_length) return;
```
Comment (behdad) [inline src/hb-font.cc]: Nope.  We still want to "reset" the coordinates.

File: src/hb-font.cc [hunk 5]
```diff
@@ -1881,6 +1882,35 @@ hb_font_set_var_coords_design (hb_font_t *font,
...
+
+  float *coords = coords_length ? (float *) calloc (coords_length, sizeof (float)) : nullptr;
+  if (unlikely (coords_length && !coords))
+    return;
+
+  hb_ot_var_named_instance_get_design_coords (font->face, instance_index, &coords_length, coords);
+  hb_font_set_var_coords_design (font, coords, coords_length);
+  if (coords)
```
Comment (behdad) [inline src/hb-font.cc]: No null check needed for `free`.

File: src/hb-font.cc (hunk 4 above)
Comment (ebraminio) [reply src/hb-font.cc]: Ah obviously, ok

**Comment by behdad** [comment]
@matthiasclasen

**Comment by matthiasclasen** [comment]
Thanks, useful

**Pull request merged**